tests/abi/functions.h

```diff
@@ -0,0 +1,309 @@
+#pragma once
+
+/*
+ * This file is distributed under the MIT License. See LICENSE.md for details.
+ */
+
+#include <assert.h>
+#include <stdint.h>
+#include <stdio.h>
+
+#include "for_each.h"
+
+/*
+ * Helper macros
+ */
+
+#ifndef ABIDEF
+#define ABIDEF
+#endif
+
+#define STRINGIFY_HELPER(X) #X
+#define STRINGIFY(X) STRINGIFY_HELPER(X)
+
+#define CONCATENATE_HELPER(LHS, RHS) LHS##RHS
+#define CONCATENATE(LHS, RHS) CONCATENATE_HELPER(LHS, RHS)
+
+#define CONCATENATE_THREE_HELPER(F, S, T) F##S##T
+#define CONCATENATE_THREE(F, S, T) CONCATENATE_THREE_HELPER(F, S, T)
+
+#define DO_NOTHING_HELPER(LHS, RHS) LHS RHS
+#define DO_NOTHING(LHS, RHS) DO_NOTHING_HELPER(LHS, RHS)
+
+static_assert(__BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+                || __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__,
+              "Unsupported endianness (pdp? something crazy? Memory "
+              "corruption?)");
+
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define PRINT_FUNCTION_HEADER(FUNCTION_NAME)            \
+  puts("  Function: \"" STRINGIFY(FUNCTION_NAME) "\"\n" \
+       "  IsLittleEndian: yes");
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#define PRINT_FUNCTION_HEADER(FUNCTION_NAME)            \
+  puts("  Function: \"" STRINGIFY(FUNCTION_NAME) "\"\n" \
+       "  IsLittleEndian: no");
+#endif
+
+#define UNROLL_ARGUMENT(LHS, RHS) DO_NOTHING(LHS, argument_##RHS)
+
+size_t runtime_endianness_check() {
+  size_t i = 1;
+  return !*((char *) &i);
+}
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#define PRINT_BYTES(VARIABLE_NAME, BYTE_COUNT)           \
+  for (size_t i = BYTE_COUNT - 1; i != (size_t) -1; --i) \
+    printf("%.2x", VARIABLE_NAME[i]);
+#elif __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#define PRINT_BYTES(VARIABLE_NAME, BYTE_COUNT) \
+  for (size_t i = 0; i < BYTE_COUNT; ++i)      \
+    printf("%.2x", VARIABLE_NAME[i]);
+#endif
+
+static_assert(sizeof(uint8_t) == 1, "A type with size == 1 is required.");
+#define PRINT_VALUE(TYPE, VALUE)         \
+  do {                                   \
+    typedef union {                      \
+      TYPE v;                            \
+      uint8_t a[sizeof(TYPE)];           \
+    } printing_helper;                   \
+                                         \
+    printing_helper helper;              \
+    helper.v = VALUE;                    \
+                                         \
+    PRINT_BYTES(helper.a, sizeof(TYPE)); \
```

Comment:
Why handle the endianess by hand when you could just `printf("%x")`?

Reply:
To support objects of arbitrary size without worrying about splitting them into register-sized burst and/or system-dependent endianness conversion methods.

Reply:
OK, but here, correct me if I'm wrong, you're using this function to print the value contained in a register, which is either 32-bit or 64-bits. `printf` would have the same effect, with less logic. No need to perform any conversion.

Reply:
I have exactly one printing "macro" which is used for everything. Be it a register or a 80-byte struct. And it's this `PRINT_BYTES` thing.

Reply:
This is gone (yay!!)